File: perudo.py
import random
import math
MAX_PLAYERS = 4
DICES_PER_PLAYER = 5
DOUBT_ACTION = (-1, -1)
JOLLY_FACE = 1

'''
RL Agent that learns how to play Perudo.
State space: 
        - total number of dices in game
        - player's own dices (as a list of counts for each face)
        - last bid made (quantity and face)
Action space:
        - make a bid (quantity and face) with constraints
        - doubt the last bid

Reward structure:
        - +1 for winning a game
        - 0 otherwise
'''

# Generate all possible dice face counts for a given number of dices
all_possible_dice_counts_table = {0 : [[0] * 6]}

for tot in range(1, DICES_PER_PLAYER + 1):
    for prev in all_possible_dice_counts_table[tot - 1]:
        for n in range(6):
            fcs = prev.copy()
            fcs[n] += 1
            all_possible_dice_counts_table[tot] = all_possible_dice_counts_table.get(tot, []) + [fcs]


        

def generate_state_space(n_players=MAX_PLAYERS):
    state_space = {}
    for tot_dices in range(2, n_players * DICES_PER_PLAYER + 1):
        cur_state = [0] * (1 + 2 + 6) # tot_dices + last bid (quantity, face) + own dices (6 faces) 
        cur_state[0] = tot_dices
        for n_dices in range(1, 6):
            for dice_confs in all_possible_dice_counts_table.get(n_dices, []):
                cur_state[3:] = dice_confs
                for last_bid_quantity in range(0, tot_dices + 1):
                    for last_bid_face in range(0, 7):
                        cur_state[1] = last_bid_quantity
                        cur_state[2] = last_bid_face
                        state_space[tuple(cur_state)] = 0
    return state_space
        
def possible_actions_from_bid(tot_dices, last_bid_quantity, last_bid_face):
        
        actions = set()
        
        min_next_bid_q = max(1, last_bid_quantity)
        min_next_bid_f = max(1, last_bid_face)
        # doubt action only if there was a last bid
        if last_bid_quantity > 0:
            actions.add(DOUBT_ACTION)

        # last bid was jolly bid?
        if last_bid_face == JOLLY_FACE:
            min_next_bid_q = 2*last_bid_quantity + 1
            min_next_bid_f = 2

        # special jolly bid            
        for bid_quantity in range((min_next_bid_q // 2) + (min_next_bid_q % 2), tot_dices + 1):
            actions.add((bid_quantity, JOLLY_FACE))

        # possible bids
        for bid_quantity in range(min_next_bid_q, tot_dices + 1):
            for bid_face in range(min_next_bid_f, 7):
                if bid_quantity == min_next_bid_q and bid_face == last_bid_face:
                    continue
                actions.add((bid_quantity, bid_face))
        return list(actions)

def generate_action_space(state_space):
    action_space = {}
    for state in state_space.keys():
        tot_dices = state[0]
        last_bid_quantity = state[1]
        last_bid_face = state[2]
        # Skip if already generated
        if (action_space.get((tot_dices, last_bid_quantity, last_bid_face))):
            continue
        actions = possible_actions_from_bid(tot_dices, last_bid_quantity, last_bid_face)
        action_space[(tot_dices, last_bid_quantity, last_bid_face)] = actions
    return action_space

def generate_policy_map(state_space, action_space):
    policy_map = {}
    for state in state_space.keys():
        actions = action_space[(state[0], state[1], state[2])]
        for a in actions:
            policy_map[state, a] = 1 # Equal weight per action
    return policy_map

''' Generate all possible states, actions and initialize policy map 
    NOT NEEDED
'''
#STATE_SPACE = generate_state_space()
#ACTION_SPACE = generate_action_space(STATE_SPACE)
#POLICY_MAP = generate_policy_map(STATE_SPACE, ACTION_SPACE)
POLICY_MAP = {}

def policy(state, epsilon=0.05):
    actions = possible_actions_from_bid(state[0], state[1], state[2])
    if random.random() < epsilon:
        return random.choice(actions)
    else:
        # Choose action with highest weight
        max_ = -math.inf
        best_action = None
        for a in actions:
            action_value = POLICY_MAP.get((state, a), 1)
            if action_value > max_:
                max_ = action_value
                best_action = a

        return best_action
        
        


def render_dices(dices):
    for idx, count in enumerate(dices):
        if count > 0:
            def f2c(f):
                return "Jolly" if f == JOLLY_FACE else str(f)
            print(f"    {count} of {f2c(idx + 1)}")

def is_legal_bid(last_bid, new_bid):
    last_bid_quantity, last_bid_face = last_bid
    new_bid_quantity, new_bid_face = new_bid

    if last_bid_quantity == 0 and last_bid_face == 0:
        return True

    if new_bid == DOUBT_ACTION:
        return last_bid_quantity > 0

    min_bid_q = last_bid_quantity
    min_bid_f = last_bid_face
    if last_bid_face == JOLLY_FACE:
        min_bid_q = 2*last_bid_quantity
        min_bid_f = 1

    if new_bid_quantity > min_bid_q and new_bid_face >= min_bid_f:
        return True
    
    if new_bid_quantity >= min_bid_q and new_bid_face > min_bid_f:
        return True


    if new_bid_face == JOLLY_FACE and new_bid_quantity >= ((min_bid_q % 2) + (min_bid_q // 2)):
        return True

    return False

class Player():
    def __init__(self, name, n_dices=DICES_PER_PLAYER):
        self.name = name
        self.n_dices = n_dices
        self.history = []
        self.new_round()
    
    def new_round(self):
        self.dices = random.choice(all_possible_dice_counts_table[self.n_dices])

    def lose_dice(self):
        self.n_dices = max(0, self.n_dices - 1)
        self.flush_history(-1)
        if self.n_dices == 0:
            self.lose()
    
    def flush_history(self, final_reward):
        reward = final_reward
        for state, action in self.history[::-1]:
            cur = POLICY_MAP.get((state, action), 1)
            POLICY_MAP[state, action] = cur + 0.1 * (reward - cur)
            reward = POLICY_MAP[state, action]

    def lose(self):
        self.flush_history(-10)

    def win(self):
        self.flush_history(10)

    def is_alive(self):
        return self.n_dices > 0
    
    def track_action(self, total_dices, last_bid, action):
        state = [0] * (1 + 2 + 6)
        state[0] = total_dices
        state[1] = last_bid[0]
        state[2] = last_bid[1]
        state[3:] = self.dices
        state = tuple(state)
        self.history.append((state, action))
    
    def make_action(self, *args):
        pass

class RLPlayer(Player):
    GAMES = 0
    WINS = 0

    @staticmethod
    def get_stats():
        return RLPlayer.WINS / RLPlayer.GAMES if RLPlayer.GAMES > 0 else 0
    
    def track_stats(self):
        self.track_stats_ = True
        RLPlayer.GAMES += 1

    def make_action(self, total_dices, last_bid):
        state = [0] * (1 + 2 + 6)
        state[0] = total_dices
        state[1] = last_bid[0]
        state[2] = last_bid[1]
        state[3:] = self.dices
        state = tuple(state)
        action = policy(state)
        return action
    
    def win(self):
        super().win()
        if self.track_stats_:
            RLPlayer.WINS += 1

    def __init__(self, name, n_dices=DICES_PER_PLAYER):
        super().__init__(name, n_dices)
        self.track_stats_ = False
        
'''
Player that uses a static strategy:
- other_dices = total_dices - self.n_dices
- estimate for every face = own_dices[face - 1] + own_dices[0] + other_dices / 3 
- estimate for jolly = own_dices[0] + other_dices / 6
- find if there is a possible action that is lower than estimate, pick the highest one
- else, doubt
'''
class RoboPlayer(Player):
    def new_round(self):
        super().new_round()
        #print(f"{self.name} dices: ")
        #render_dices(self.dices)

    def get_estimate_vector(self, total_dices):
        estimate = [0] * 6
        other_dices = total_dices - self.n_dices
        for face in range(1, 7):
            if face == JOLLY_FACE:
                estimate[face - 1] = self.dices[0] + other_dices / 6
            else:
                estimate[face - 1] = self.dices[face - 1] + self.dices[0] + other_dices / 3
        return estimate
    
    def aggressive_action(self, total_dices, last_bid):
        '''
        Always make the highest possible bid under estimate
        '''
        estimate = self.get_estimate_vector(total_dices)

        last_bid_quantity, last_bid_face = last_bid
        possible_actions = possible_actions_from_bid(total_dices, last_bid_quantity, last_bid_face)
        best_action = DOUBT_ACTION
        best_quantity = -1

        for action in possible_actions:
            bid_quantity, bid_face = action
            if action == DOUBT_ACTION:
                continue
            if bid_quantity <= estimate[bid_face - 1]:
                if bid_quantity > best_quantity:
                    best_quantity = bid_quantity
                    best_action = action

        return best_action

    def conservative_action(self, total_dices, last_bid):
        '''
        Always make the lowest possible bid under estimate
        '''
        estimate = self.get_estimate_vector(total_dices)

        last_bid_quantity, last_bid_face = last_bid
        possible_actions = possible_actions_from_bid(total_dices, last_bid_quantity, last_bid_face)
        best_action = DOUBT_ACTION
        best_quantity = total_dices + 1

        for action in possible_actions:
            bid_quantity, bid_face = action
            if action == DOUBT_ACTION:
                continue
            if bid_quantity <= estimate[bid_face - 1]:
                if bid_quantity < best_quantity:
                    best_quantity = bid_quantity
                    best_action = action

        return best_action

    def make_action(self, total_dices, last_bid):
        pass

class AggressiveRoboPlayer(RoboPlayer):
    def make_action(self, total_dices, last_bid):
        return self.aggressive_action(total_dices, last_bid)

class ConservativeRoboPlayer(RoboPlayer):
    def make_action(self, total_dices, last_bid):
        return self.conservative_action(total_dices, last_bid)
    
class HumanPlayer(Player):
    def make_action(self, total_dices, last_bid):
        print(f"Your dices: ")
        render_dices(self.dices)
        print(f"Total dices in game: {total_dices}")
        print(f"Last bid: {last_bid}")
        while True:
            print("Bid quantity: (-1 for doubt)")
            bid_quantity = int(input())
            if bid_quantity == -1:
                return DOUBT_ACTION
            print("Bid face (1-6): ")
            bid_face = int(input())
            if is_legal_bid(last_bid, (bid_quantity, bid_face)):
                return (bid_quantity, bid_face)
            print("Illegal bid, try again.")

class Game():
    STATS = {}
    def __init__(self, players, quiet=False):
        self.players = players
        self.n_players = len(players)
        self.start_idx = 0
        self.quiet = quiet
    
    def print(self, str_):
        if self.quiet:
            return
        print(str_)

    def play_round(self):
        self.print("---------------------")
        self.print("Starting a new round...")
        for p in self.players:
            p.new_round()
        
        total_dices = sum([p.n_dices for p in self.players])
        last_bid = (0, 0)
        current_player_idx = self.start_idx % self.n_players

        while True:

            current_player = self.players[current_player_idx]
            action = current_player.make_action(total_dices, last_bid)
            current_player.track_action(total_dices, last_bid, action)
            self.print(f"Player {current_player_idx}, {current_player.name} action: {action}")

            if action == DOUBT_ACTION:
                # Resolve doubt
                bidder_idx = (current_player_idx - 1) % self.n_players
                bidder = self.players[bidder_idx]
                actual_count = 0
                q, f = last_bid
                for p in self.players:
                    actual_count += p.dices[0]
                    if f != JOLLY_FACE:
                        actual_count += p.dices[f - 1]
                
                self.print(f"Actual count for face {f}: {actual_count}")

                if actual_count >= q:
                    # Bidder wins
                    self.start_idx = current_player_idx
                    current_player.lose_dice()
                    if (not current_player.is_alive()):
                        self.print(f"Player {current_player_idx} has been eliminated!")
                        self.players.pop(current_player_idx)
                        self.n_players -= 1
                else:
                    # Doubter wins
                    self.start_idx = bidder_idx
                    bidder.lose_dice()
                    if (not bidder.is_alive()):
                        self.print(f"Player {bidder_idx} has been eliminated!")
                        self.players.pop(bidder_idx)
                        self.n_players -= 1
                break
            else:
                last_bid = action
                current_player_idx = (current_player_idx + 1) % self.n_players
        
        if self.n_players == 1:
            winner = self.players[0]
            winner.win()
            self.print(f"We have a winner: {winner.name}")
            Game.STATS[winner.name] = Game.STATS.get(winner.name, 0) + 1
            return True
        
        return False

    def play_game(self, with_human=False):
        game_over = False
        while not game_over:
            game_over = self.play_round()
            if with_human:
                print("Press Enter to continue to next round...")
                input()

def train_against_random_models(n=10000, n_players=MAX_PLAYERS):
    models = [AggressiveRoboPlayer, ConservativeRoboPlayer, RLPlayer]
    others = random.choices(models, k=n_players - 1)
    return trainRL(n=n, opponent_models=others)

def trainRL(n=10000, n_players=MAX_PLAYERS, opponent_models=[AggressiveRoboPlayer, ConservativeRoboPlayer]):

    # Reset RLPlayer stats
    RLPlayer.GAMES = 0
    RLPlayer.WINS = 0
    
    for _ in range(n):
        rlplayer = RLPlayer("RL-Agent")
        rlplayer.track_stats()
        players = [rlplayer] + [m(f"Player-{i}") for i, m in enumerate(opponent_models[:n_players - 1])]
        game = Game(players, quiet=True)
        game.play_game()
    
    return RLPlayer.get_stats()



def main():
    print("Play against the trained RL agent!")
    
    game = Game([HumanPlayer("Human"), RLPlayer("RL1")])
    game.play_game(True)

File: perudo.js
// Constants
const MAX_PLAYERS = 4;
const DICES_PER_PLAYER = 5;
const DOUBT_ACTION = [-1, -1];
const JOLLY_FACE = 1;

// Generate all possible dice face counts for a given number of dices
const allPossibleDiceCountsTable = {};

function generateDiceCounts() {
    allPossibleDiceCountsTable[0] = [[0, 0, 0, 0, 0, 0]];
    
    for (let tot = 1; tot <= DICES_PER_PLAYER; tot++) {
        allPossibleDiceCountsTable[tot] = [];
        if (tot === 1) {
            for (let n = 0; n < 6; n++) {
                const fcs = [0, 0, 0, 0, 0, 0];
                fcs[n] = 1;
                allPossibleDiceCountsTable[tot].push(fcs);
            }
        } else {
            for (const prev of allPossibleDiceCountsTable[tot - 1]) {
                for (let n = 0; n < 6; n++) {
                    const fcs = [...prev];
                    fcs[n] += 1;
                    allPossibleDiceCountsTable[tot].push(fcs);
                }
            }
        }
    }
}

generateDiceCounts();

// Helper function to check if two arrays are equal
function arraysEqual(a, b) {
    return a.length === b.length && a.every((val, idx) => val === b[idx]);
}

// Check if a bid is legal
function isLegalBid(lastBid, newBid) {
    const [lastBidQuantity, lastBidFace] = lastBid;
    const [newBidQuantity, newBidFace] = newBid;

    if (lastBidQuantity === 0 && lastBidFace === 0) {
        return true;
    }

    if (arraysEqual(newBid, DOUBT_ACTION)) {
        return lastBidQuantity > 0;
    }

    let minBidQ = lastBidQuantity;
    let minBidF = lastBidFace;
    
    if (lastBidFace === JOLLY_FACE) {
        minBidQ = 2 * lastBidQuantity + 1;
        minBidF = 1;
    }

    if (newBidQuantity > minBidQ && newBidFace >= minBidF) {
        return true;
    }
    
    if (newBidQuantity >= minBidQ && newBidFace > minBidF) {
        return true;
    }

    if (newBidFace === JOLLY_FACE && newBidQuantity >= ((minBidQ % 2) + Math.floor(minBidQ / 2))) {
        return true;
    }

    return false;
}

// Get possible actions from a bid
function possibleActionsFromBid(totDices, lastBidQuantity, lastBidFace) {
    const actions = new Set();
    
    let minNextBidQ = Math.max(1, lastBidQuantity);
    let minNextBidF = Math.max(2, lastBidFace);
    
    // Doubt action only if there was a last bid
    if (lastBidQuantity > 0) {
        actions.add(JSON.stringify(DOUBT_ACTION));
    }

    // Last bid was jolly bid?
    if (lastBidFace === JOLLY_FACE) {
        minNextBidQ = 2 * lastBidQuantity + 1;
        minNextBidF = 2;
    }

    // Special jolly bid (only from second bid onwards)
    if (lastBidFace !== 0) {
        for (let bidQuantity = Math.floor(minNextBidQ / 2) + (minNextBidQ % 2); bidQuantity <= totDices; bidQuantity++) {
            actions.add(JSON.stringify([bidQuantity, JOLLY_FACE]));
        }
    }

    // Possible bids
    for (let bidQuantity = minNextBidQ; bidQuantity <= totDices; bidQuantity++) {
        for (let bidFace = minNextBidF; bidFace <= 6; bidFace++) {
            if (bidQuantity === minNextBidQ && bidFace === lastBidFace) {
                continue;
            }
            actions.add(JSON.stringify([bidQuantity, bidFace]));
        }
    }
    
    return Array.from(actions).map(a => JSON.parse(a));
}

// Player base class
class Player {
    constructor(name, nDices = DICES_PER_PLAYER) {
        this.name = name;
        this.nDices = nDices;
        this.history = [];
        this.newRound();
    }

    newRound() {
        const possibleCounts = allPossibleDiceCountsTable[this.nDices];
        this.dices = possibleCounts[Math.floor(Math.random() * possibleCounts.length)];
    }

    winBid() {
        // Small reward for winning one bid
    }

    loseDice() {
        this.nDices = Math.max(0, this.nDices - 1);
        if (this.nDices === 0) {
            this.lose();
        }
    }

    lose() {
        // Player lost
    }

    win() {
        // Player won
    }

    isAlive() {
        return this.nDices > 0;
    }

    makeAction(totalDices, lastBid) {
        // To be implemented by subclasses
        return null;
    }
}

// Human Player
class HumanPlayer extends Player {
    constructor(name, nDices = DICES_PER_PLAYER) {
        super(name, nDices);
        this.pendingAction = null;
        this.actionResolve = null;
    }

    makeAction(totalDices, lastBid) {
        return new Promise((resolve) => {
            this.actionResolve = resolve;
            // UI will call setAction when user makes a choice
        });
    }

    setAction(action) {
        if (this.actionResolve) {
            this.actionResolve(action);
            this.actionResolve = null;
        }
    }
}

// RoboPlayer base class
class RoboPlayer extends Player {
    getEstimateVector(totalDices) {
        const estimate = [0, 0, 0, 0, 0, 0];
        const otherDices = totalDices - this.nDices;
        
        for (let face = 1; face <= 6; face++) {
            if (face === JOLLY_FACE) {
                estimate[face - 1] = this.dices[0] + otherDices / 6;
            } else {
                estimate[face - 1] = this.dices[face - 1] + this.dices[0] + otherDices / 3;
            }
        }
        return estimate;
    }

    aggressiveAction(totalDices, lastBid) {
        const estimate = this.getEstimateVector(totalDices);
        const [lastBidQuantity, lastBidFace] = lastBid;
        const possibleActions = possibleActionsFromBid(totalDices, lastBidQuantity, lastBidFace);
        let bestAction = DOUBT_ACTION;
        let bestQuantity = -1;

        for (const action of possibleActions) {
            const [bidQuantity, bidFace] = action;
            if (arraysEqual(action, DOUBT_ACTION)) {
                continue;
            }
            if (bidQuantity <= estimate[bidFace - 1]) {
                if (bidQuantity > bestQuantity) {
                    bestQuantity = bidQuantity;
                    bestAction = action;
                }
            }
        }

        return bestAction;
    }

    conservativeAction(totalDices, lastBid) {
        const estimate = this.getEstimateVector(totalDices);
        const [lastBidQuantity, lastBidFace] = lastBid;
        const possibleActions = possibleActionsFromBid(totalDices, lastBidQuantity, lastBidFace);
        let bestAction = DOUBT_ACTION;
        let bestQuantity = totalDices + 1;

        for (const action of possibleActions) {
            const [bidQuantity, bidFace] = action;
            if (arraysEqual(action, DOUBT_ACTION)) {
                continue;
            }
            if (bidQuantity <= estimate[bidFace - 1]) {
                if (bidQuantity < bestQuantity) {
                    bestQuantity = bidQuantity;
                    bestAction = action;
                }
            }
        }

        return bestAction;
    }

    safeMinimalBid(totalDices, lastBid) {
        const estimate = this.getEstimateVector(this.nDices);
        const [lastBidQuantity, lastBidFace] = lastBid;
        const possibleActions = possibleActionsFromBid(totalDices, lastBidQuantity, lastBidFace);
        let bestAction = DOUBT_ACTION;
        let bestQuantity = totalDices + 1;

        for (const action of possibleActions) {
            const [bidQuantity, bidFace] = action;
            if (arraysEqual(action, DOUBT_ACTION)) {
                continue;
            }
            if (bidQuantity <= estimate[bidFace - 1]) {
                if (bidQuantity < bestQuantity) {
                    bestQuantity = bidQuantity;
                    bestAction = action;
                }
            }
        }

        return bestAction;
    }
}

// RoboPlayer variants
class AggressiveRoboPlayer extends RoboPlayer {
    makeAction(totalDices, lastBid) {
        return this.aggressiveAction(totalDices, lastBid);
    }
}

class ConservativeRoboPlayer extends RoboPlayer {
    makeAction(totalDices, lastBid) {
        return this.conservativeAction(totalDices, lastBid);
    }
}

class DoubterRoboPlayer extends RoboPlayer {
    makeAction(totalDices, lastBid) {
        return this.safeMinimalBid(totalDices, lastBid);
    }
}

// Game class
class Game {
    constructor(players) {
        this.players = players;
        this.nPlayers = players.length;
        this.startIdx = 0;
        this.currentRound = 0;
        this.lastBid = [0, 0];
        this.currentPlayerIdx = 0;
    }

    async playRound() {
        this.currentRound++;
        console.log("Starting a new round...");
        
        for (const p of this.players) {
            p.newRound();
        }
        
        const totalDices = this.players.reduce((sum, p) => sum + p.nDices, 0);
        this.lastBid = [0, 0];
        this.currentPlayerIdx = this.startIdx % this.nPlayers;

        while (true) {
            const currentPlayer = this.players[this.currentPlayerIdx];
            updateUI(this.currentPlayerIdx, this.lastBid, totalDices);
            
            let action;
            if (currentPlayer instanceof HumanPlayer) {
                action = await currentPlayer.makeAction(totalDices, this.lastBid);
            } else {
                // Add delay for robot players
                await new Promise(resolve => setTimeout(resolve, 1000));
                action = currentPlayer.makeAction(totalDices, this.lastBid);
            }

            console.log(`Player ${this.currentPlayerIdx}, ${currentPlayer.name} action:`, action);

            if (arraysEqual(action, DOUBT_ACTION)) {
                // Resolve doubt
                const bidderIdx = (this.currentPlayerIdx - 1 + this.nPlayers) % this.nPlayers;
                const bidder = this.players[bidderIdx];
                let actualCount = 0;
                const [q, f] = this.lastBid;
                
                for (const p of this.players) {
                    actualCount += p.dices[0]; // Jollies
                    if (f !== JOLLY_FACE) {
                        actualCount += p.dices[f - 1];
                    }
                }

                showMessage(`Actual count for face ${f === JOLLY_FACE ? 'Jolly' : f}: ${actualCount}`, 'info');

                if (actualCount >= q) {
                    // Bidder wins
                    this.startIdx = this.currentPlayerIdx;
                    currentPlayer.loseDice();
                    bidder.winBid();
                    showMessage(`${currentPlayer.name} doubted incorrectly! ${currentPlayer.name} loses a die.`, 'error');
                    
                    if (!currentPlayer.isAlive()) {
                        showMessage(`${currentPlayer.name} has been eliminated!`, 'error');
                        this.players.splice(this.currentPlayerIdx, 1);
                        this.nPlayers--;
                        // Adjust start index if needed
                        if (this.startIdx >= this.nPlayers) {
                            this.startIdx = 0;
                        }
                    }
                } else {
                    // Doubter wins
                    this.startIdx = bidderIdx;
                    bidder.loseDice();
                    currentPlayer.winBid();
                    showMessage(`${bidder.name} bid incorrectly! ${bidder.name} loses a die.`, 'success');
                    
                    if (!bidder.isAlive()) {
                        showMessage(`${bidder.name} has been eliminated!`, 'error');
                        const removeIdx = this.players.indexOf(bidder);
                        this.players.splice(removeIdx, 1);
                        this.nPlayers--;
                        // Adjust indices if needed
                        if (removeIdx < this.currentPlayerIdx) {
                            this.currentPlayerIdx--;
                        }
                        if (this.startIdx >= this.nPlayers) {
                            this.startIdx = 0;
                        }
                    }
                }
                break;
            } else {
                this.lastBid = action;
                const [q, f] = action;
                showMessage(`${currentPlayer.name} bids: ${q} ${f === JOLLY_FACE ? 'Jolly' : f}`, 'info');
                this.currentPlayerIdx = (this.currentPlayerIdx + 1) % this.nPlayers;
            }
        }
        
        if (this.nPlayers === 1) {
            const winner = this.players[0];
            winner.win();
            showMessage(`We have a winner: ${winner.name}!`, 'success');
            document.getElementById('gameEndActions').classList.remove('hidden');
            return true;
        }
        
        return false;
    }

    async playGame() {
        let gameOver = false;
        while (!gameOver) {
            gameOver = await this.playRound();
            if (!gameOver) {
                await new Promise(resolve => setTimeout(resolve, 2000));
            }
        }
    }
}

// Global game instance
let game = null;
let humanPlayer = null;

// UI Functions
function renderDice(diceValue) {
    const dice = document.createElement('div');
    dice.className = 'dice';
    
    if (diceValue === 1) {
        dice.classList.add('jolly');
        dice.textContent = '🎲';
    } else {
        const face = document.createElement('div');
        face.className = 'dice-face';
        face.style.display = 'grid';
        face.style.gridTemplateColumns = 'repeat(3, 1fr)';
        face.style.gridTemplateRows = 'repeat(3, 1fr)';
        face.style.width = '100%';
        face.style.height = '100%';
        face.style.padding = '5px';
        
        // Create dots based on dice value
        const dotPositions = {
            2: [[0,0], [2,2]],
            3: [[0,0], [1,1], [2,2]],
            4: [[0,0], [0,2], [2,0], [2,2]],
            5: [[0,0], [0,2], [1,1], [2,0], [2,2]],
            6: [[0,0], [0,1], [0,2], [2,0], [2,1], [2,2]]
        };
        
        const positions = dotPositions[diceValue] || [];
        
        for (let row = 0; row < 3; row++) {
            for (let col = 0; col < 3; col++) {
                const cell = document.createElement('div');
                cell.style.display = 'flex';
                cell.style.alignItems = 'center';
                cell.style.justifyContent = 'center';
                
                const hasDot = positions.some(p => p[0] === row && p[1] === col);
                if (hasDot) {
                    const dot = document.createElement('div');
                    dot.className = 'dot';
                    cell.appendChild(dot);
                }
                
                face.appendChild(cell);
            }
        }
        
        dice.appendChild(face);
    }
    
    return dice;
}

function renderPlayerDice(dices) {
    const container = document.getElementById('playerDice');
    container.innerHTML = '';
    
    for (let face = 0; face < 6; face++) {
        for (let count = 0; count < dices[face]; count++) {
            container.appendChild(renderDice(face + 1));
        }
    }
}

function updatePlayerStatus() {
    const statusDiv = document.getElementById('playerStatus');
    statusDiv.innerHTML = '';
    
    if (!game) return;
    
    game.players.forEach((player, idx) => {
        const playerDiv = document.createElement('div');
        playerDiv.className = 'player-status';
        if (idx === game.currentPlayerIdx) {
            playerDiv.classList.add('current');
        }
        playerDiv.textContent = `${player.name}: ${player.nDices} dice`;
        statusDiv.appendChild(playerDiv);
    });
}

function updateUI(currentPlayerIdx, lastBid, totalDices) {
    if (!game) return;
    
    updatePlayerStatus();
    
    // Always show player dice
    if (humanPlayer) {
        renderPlayerDice(humanPlayer.dices);
    }
    
    // Update last bid display
    const [q, f] = lastBid;
    if (q === 0 && f === 0) {
        document.getElementById('lastBid').textContent = 'No bid yet';
    } else {
        document.getElementById('lastBid').textContent = `Last bid: ${q} ${f === JOLLY_FACE ? 'Jolly' : f}`;
    }
    
    // Show bid section only for human player's turn
    if (currentPlayerIdx === 0 && humanPlayer) {
        document.getElementById('bidSection').classList.remove('hidden');
        document.getElementById('waitingMessage').classList.add('hidden');
        document.getElementById('doubtButton').disabled = (q === 0 && f === 0);
    } else {
        document.getElementById('bidSection').classList.add('hidden');
        document.getElementById('waitingMessage').classList.remove('hidden');
    }
}

function showMessage(text, type) {
    const messageArea = document.getElementById('messageArea');
    const message = document.createElement('div');
    message.className = `message ${type}`;
    message.textContent = text;
    messageArea.innerHTML = '';
    messageArea.appendChild(message);
}

function makeBid() {
    if (!humanPlayer || !humanPlayer.actionResolve) return;
    if (!game) return;
    
    const quantity = parseInt(document.getElementById('bidQuantity').value);
    const face = parseInt(document.getElementById('bidFace').value);
    
    if (isNaN(quantity) || quantity < 1) {
        showMessage('Please enter a valid quantity', 'error');
        return;
    }
    
    const newBid = [quantity, face];
    const actualLastBid = game.lastBid || [0, 0];
    
    if (!isLegalBid(actualLastBid, newBid)) {
        showMessage('Illegal bid! Please check the rules.', 'error');
        return;
    }
    
    humanPlayer.setAction(newBid);
    document.getElementById('bidQuantity').value = '';
}

function doubtBid() {
    if (!humanPlayer || !humanPlayer.actionResolve) return;
    
    const lastBidText = document.getElementById('lastBid').textContent;
    if (lastBidText === 'No bid yet') {
        showMessage('Cannot doubt when there is no bid', 'error');
        return;
    }
    
    humanPlayer.setAction(DOUBT_ACTION);
    showMessage('You doubt the last bid!', 'info');
}

function resetGame() {
    document.getElementById('gameEndActions').classList.add('hidden');
    document.getElementById('gameSetup').classList.remove('hidden');
    document.getElementById('gameArea').classList.add('hidden');
    game = null;
    humanPlayer = null;
}

function startNewGame() {
    const opponentType = document.getElementById('opponentSelect').value;
    
    humanPlayer = new HumanPlayer('You');
    let opponent;
    
    switch (opponentType) {
        case 'aggressive':
            opponent = new AggressiveRoboPlayer('Aggressive Bot');
            break;
        case 'conservative':
            opponent = new ConservativeRoboPlayer('Conservative Bot');
            break;
        case 'doubter':
            opponent = new DoubterRoboPlayer('Doubter Bot');
            break;
    }
    
    game = new Game([humanPlayer, opponent]);
    
    document.getElementById('gameSetup').classList.add('hidden');
    document.getElementById('gameArea').classList.remove('hidden');
    document.getElementById('gameEndActions').classList.add('hidden');
    document.getElementById('messageArea').innerHTML = '';
    
    game.playGame();
}

// Allow Enter key to submit bid
document.addEventListener('DOMContentLoaded', () => {
    document.getElementById('bidQuantity').addEventListener('keypress', (e) => {
        if (e.key === 'Enter') {
            makeBid();
        }
    });
});
